src/mlpack/base.hpp

```diff
@@ -0,0 +1,124 @@
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_BASE_HPP
+#define MLPACK_BASE_HPP
+
+// First, check if Armadillo was included before, warning if so.
+#ifdef ARMA_INCLUDES
+#pragma message "Armadillo was included before mlpack; this can sometimes cause\
+ problems.  It should only be necessary to include <mlpack/core.hpp> and not \
+<armadillo>."
+#endif
+
+// Defining _USE_MATH_DEFINES should set M_PI.
+#define _USE_MATH_DEFINES
+#include <cmath>
+
+// Next, standard includes.
+#include <cctype>
+#include <cfloat>
+#include <climits>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <stdexcept>
+#include <tuple>
+#include <utility>
+#include <numeric>
+#include <vector>
+#include <queue>
+
+// But if it's not defined, we'll do it.
+#ifndef M_PI
+  #define M_PI 3.141592653589793238462643383279
+#endif
+
+// MLPACK_COUT_STREAM is used to change the default stream for printing
+// purpose.
+#if !defined(MLPACK_COUT_STREAM)
+ #define MLPACK_COUT_STREAM std::cout
+#endif
+
+// MLPACK_CERR_STREAM is used to change the stream for printing warnings
+// and errors.
+#if !defined(MLPACK_CERR_STREAM)
+ #define MLPACK_CERR_STREAM std::cerr
+#endif
+
+// Give ourselves a nice way to force functions to be inline if we need.
+#define force_inline
+#if defined(__GNUG__) && !defined(DEBUG)
+  #undef force_inline
+  #define force_inline __attribute__((always_inline))
+#elif defined(_MSC_VER) && !defined(DEBUG)
+  #undef force_inline
+  #define force_inline __forceinline
+#endif
+
+// Backport this functionality from C++14, if it doesn't exist.
+#if __cplusplus <= 201103L
+#if !defined(_MSC_VER) || _MSC_VER <= 1800
+namespace std {
+
+template<bool B, class T = void>
+using enable_if_t = typename enable_if<B, T>::type;
+
+}
+#endif
+#endif
+
+// Backport std::any from C+17 to C++11 to replace boost::any.
+// Use mnmlstc backport implementation only if compiler does not
+// support C++17.
+#if __cplusplus < 201703L
+  #include <mlpack/core/std_backport/any.hpp>
+  #include <mlpack/core/std_backport/string_view.hpp>
+  #define ANY core::v2::any
+  #define ANY_CAST core::v2::any_cast
+  #define STRING_VIEW core::v2::string_view
+#else
+  #include <any>
+  #include <string_view>
+  #define ANY std::any
+  #define ANY_CAST std::any_cast
+  #define STRING_VIEW std::string_view
+#endif 
+
+// Now include Armadillo through the special mlpack extensions.
+#include <mlpack/core/arma_extend/arma_extend.hpp>
+#include <mlpack/core/util/arma_traits.hpp>
+
+// On Visual Studio, disable C4519 (default arguments for function templates)
+// since it's by default an error, which doesn't even make any sense because
+// it's part of the C++11 standard.
+#ifdef _MSC_VER
+  #pragma warning(disable : 4519)
+  #define ARMA_USE_CXX11
```

Comment:
Not sure if `#define ARMA_USE_CXX11` is necessary.  Armadillo versions >= 10 require C++11 as the minimum standard.  Armadillo versions < 10 already internally enable C++11 features when a compiler capable of handling C++11 is detected; this detection includes workarounds for dodgy MSVC "compilers".

Reply:
I got curious and went digging for when this popped up.  Turns out it was [2016](https://github.com/mlpack/mlpack/commit/6f213a3d077ddde4281e072e2a975398c67a253f), more recently than I would have guessed.  Not much in the way of [comments](https://github.com/mlpack/mlpack/pull/523/files) when it was added; I guess it was not noticed and then immediately forgotten. :smile: Removed in fe73ae2cf6.